extensions/amp-social-share/0.2/social-share.js

```diff
@@ -30,63 +32,196 @@ const DEFAULT_HEIGHT = 44;
+   * @param {?string} name
+   * @return {!JsonObject}
    */
-  function handleKeyPress(event) {
-    const {key} = event;
-    if (key == Keys.SPACE || key == Keys.ENTER) {
-      event.preventDefault();
-      handleActivation();
+  function checkProps_(props, name) {
+    const {type, shareEndpoint, params, bindings, width, height} = props;
+
+    // Verify type is valid and cannot contain spaces
+    if (type === undefined) {
+      throwError_(`The type attribute is required. ${name}`);
+    }
+    if (/\s/.test(type)) {
+      throwError_(
+        `Space characters are not allowed in type attribute value. ${name}`
+      );
+    }
+
+    // bindings and params props must be objects
+    if (params && !isObject(params)) {
+      throwError_(`The params property should be an object. ${name}`);
+    }
+    if (bindings && !isObject(bindings)) {
+      throwError_(`The bindings property should be an object. ${name}`);
+    }
+
+    // User must provide shareEndpoint if they choose a type that is not
+    // pre-configured
+    const typeConfig = getSocialConfig(type) || dict();
+    const baseEndpoint = typeConfig['shareEndpoint'] || shareEndpoint;
+    if (baseEndpoint === undefined) {
+      throwError_(
+        `A shareEndpoint is required if not using a pre-configured type. ${name}`
+      );
+    }
+
+    // Throw warning if type is obsolete
+    const obsoleteType = typeConfig['obsolete'];
+    if (obsoleteType) {
+      throwWarning_(`Skipping obsolete share button ${type}. ${name}`);
     }
+
+    // Verify width and height are valid integers
+    let checkedWidth = width || DEFAULT_WIDTH;
+    let checkedHeight = height || DEFAULT_HEIGHT;
+    if (width && !Number.isInteger(width)) {
+      throwWarning_(
+        `The width property should be an Integer, defaulting to ${DEFAULT_WIDTH}. ${name}`
+      );
+      checkedWidth = DEFAULT_WIDTH;
+    }
+    if (height && !Number.isInteger(height)) {
+      throwWarning_(
+        `The height property should be an Integer, defaulting to ${DEFAULT_HEIGHT}. ${name}`
+      );
+      checkedHeight = DEFAULT_HEIGHT;
+    }
+    return {
+      typeConfig,
+      baseEndpoint,
+      checkedWidth,
+      checkedHeight,
+      obsoleteType,
+    };
   }
 
-  /** @private */
-  function handleActivation() {
-    const href = props['href'];
-    const target = props['target'] || '_blank';
-    if (!href) {
-      throw new Error('Clicked before href is set.');
+  /**
+   * @private
+   * @param {!JsonObject} typeConfig
+   * @param {?string} baseEndpoint
+   * @param {!JsonObject} props
+   * @return {?string}
+   */
+  function createEndpoint_(typeConfig, baseEndpoint, props) {
+    const {params, bindings} = props;
+    const combinedParams = dict();
+    Object.assign(combinedParams, typeConfig['defaultParams'], params);
+    const endpointWithParams = addParamsToUrl(baseEndpoint, combinedParams);
+
+    const combinedBindings = dict();
+    const bindingVars = typeConfig['bindings'];
```

Comment:
Can you described and give examples of what params are and what bindings are? So far I'm still really confused about these.

Reply:
Here is a summary of how the params and bindings work: https://docs.google.com/document/d/1mtZu3NbD4tgmXtOTtYD_lEvMFUgebEyE4ZJc1ret-Wg/edit#heading=h.ia142csj1140

Params are an object.  Each entry in the object will be appended to the shareEndpoint, usually as searchParams (dependent on the type).  So if params is
{url: "google.com", userid: "3"}, the endpoint would look like "endpoint?url=google.com&userid=3" (it will be url encoded as well)

Bindings allows variables to be used in the endpoint and params.  So if the bindings are {recipient: "testuser} and the endpoint is "endpoint?url=RECIPIENT&userid=3" then the component would process it out to "endpoint?url=testuser&userid=3"

Reply:
This section takes the logic from the current amp version of the component here: https://github.com/ampproject/amphtml/blob/master/extensions/amp-social-share/0.2/amp-social-share.js#L144

And in preact we can do it synchronously.  My thought is that the amp component can create the bindings object and pass it to the preact component, then the preact component can handle the logic of creating the endpoint with params and bindings processed.

Reply:
These are examples of the bindings right? https://github.com/ampproject/amphtml/blob/master/spec/amp-var-substitutions.md Are they useful outside of an AMP document?

Reply:
Ok, so the next step for us to figure out what the params + bindings look like in the call to the Preact component. We can always convert the bindings/params as AMP ecosystem needs them. But on the Preact level on of them seems not needed, and the other may need to be passed as props sometimes.

E.g. that specific example, a more natural call is:

```
<social-share type="twitter" url="google.com" />
```

or at least:
```
<social-share type="twitter" params={{url: 'google.com'}} />
```

I'd really like to figure out how to just do props, at this point.

And to also stress: our current config is AMP specific config. It either needs to change significantly or split into two parts: Preact-vs-AMP. E.g. the fact that we bind AMP's "url" prop to the `CANONICAL_URL`, is specifically AMP's view of things, I believe.

For instance, the config could look like this:

```
const TYPES = {
  LINKEDIN: {
    /* Props. Come up with anything that looks natural to us internally. */
    requiredProps: ['url'],
    shareEndpoint: ({url, mini = true}) => `https://www.linkedin.com/shareArticle?url=${enc(url)&mini=${enc(mini)}}`,
  },
}
```

Thus, the rendering here would be trivial:
```
const config = TYPES[type];
if (config) {
  validateProps(props, config);
  endpoint = config. shareEndpoint(props);
}
```

Reply:
That does leave us with the question of what we should do with a custom endpoint. So we just need to see what's the most natural for it. And, IMHO, we can just pass the whole complete `shareEndpoint` with all params interpolated.

Reply:
I see your point.  And we definitely can change this to accept everything through props.  The config model you described has the benefit of making the logic within the Preact component simpler by removing the binding logic, and tradeoff the following:

- Cannot add arbitrary additional parameters to the url (all params per type would need to be pre-defined in the config)
- For custom url, the user would not be able to rely on our logic to format any parameters to be appended to the end of the url (more logic required from user outside of our component)

I guess the question is, is simplicity of the component more desirable vs. more functionality.

~~~~

Regarding how the amp component will pass in the parameters, I am not sure yet how this will work.  How do you feel about this proposal?

1. Merge the current version of preact component
2. Work on amp component updates
3. Understand with regard to both preact and amp what makes the most sense for us to do
4. Create a separate PR that will include amp component and changes to preact if needed (which may include, the simplified config described above)

Reply:
It's fine to merge this version and iterate. But to answer the question of additional params - that's easy enough to add with a `additionalParams` prop. My point was that the key parameters should be top-level props. Other things can be tacked on as needed.